crates/sui-graphql-rpc/src/types/transaction_block/mod.rs

```diff
@@ -297,102 +263,114 @@ impl TransactionBlock {
     ///
     /// If the `Page<Cursor>` is set, then this function will defer to the `checkpoint_viewed_at` in
     /// the cursor if they are consistent.
+    ///
+    /// Filters that involve a combination of `recvAddress`, `inputObject`, `changedObject`, and
+    /// `function` should provide a value for `scan_limit`. This indicates how many transactions to
+    /// scan through per the filter conditions.
     pub(crate) async fn paginate(
-        db: &Db,
+        ctx: &Context<'_>,
         page: Page<Cursor>,
         filter: TransactionBlockFilter,
         checkpoint_viewed_at: u64,
+        scan_limit: Option<u64>,
     ) -> Result<Connection<String, TransactionBlock>, Error> {
-        use transactions as tx;
+        if filter.is_empty() {
+            return Ok(Connection::new(false, false));
+        }
+
+        // If there is more than one `complex_filter` specified, then the caller has provided some
+        // arbitrary combination of `function`, `kind`, `recvAddress`, `inputObject`, or
+        // `changedObject`. Consequently, we require setting a `scanLimit`, or else we will return
+        // an error.
+        if filter.complex_filters() >= 2 && scan_limit.is_none() {
```

Comment:
Note that this currently allows adding `transaction_ids` to any query without a scan limit, but note that we decided not to do that -- even though joining against `tx_digests` should be highly selective, in rare circumstances, the query planner still comes up with a bad query plan and doesn't pick the digests table as the inner one.

Eventually, we'll be able to support this by loading all the transaction IDs into memory, and filtering them there, but IIRC we're only doing that as a follow-up, so until then, we need to treat transaction IDs as a "complex filter" for these purposes.

(But -- extra curve ball -- as noted below, because `tx_digests` doesn't have a `sender` column, we shouldn't treat it as a `complex_filter` for the purposes of deciding whether to introduce an explicit `select_senders` or not).

Reply:
Ah right, thanks for the reminder. This must've slipped my mind somehow. I think we're better off with two classifications, a `complex_filter` or nomenclature that requires `scan_limit`, and another that requires explicit `select_senders` (maybe just `requires_explicit_senders`)

Reply:
Agreed!